### PR TITLE
feat: add vcluster e2e cluster provider

### PR DESCRIPTION
**Pull request by crandles**
addition of vcluster provider; interested to see if the CI fails. vcluster needs a "host" cluster to launch into, so I've followed an existing example where a real cluster or kind cluster are utilized as necessary: https://github.com/kubernetes-sigs/e2e-framework/blob/9206e6f6082907a1c77750b7cac0fc44ff4dac02/examples/real_cluster/main_test.go#L39

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR enables a new `E2EClusterProvider` for `vcluster` based clusters.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., Usage docs, etc.:

```docs
NA
```

### CODE REVIEW

**Comment by vladimirvivien** [comment]
@crandles this is awesome and thanks for the contribution.
At first glance it looks Okay, but I will ask @harshanarayana to take a look as well since he's the E2E provider expert 😉

File: third_party/vcluster/vcluster.go
Comment (vladimirvivien) [inline third_party/vcluster/vcluster.go:159]: I think @harshanarayana has some util wrapper for running procs that should be considered here.

File: third_party/vcluster/vcluster.go
Comment (vladimirvivien) [inline third_party/vcluster/vcluster.go:126]: @crandles does this vcluster create function will only work for kind "real" clusters (i don't know much about vclusters). If that is the case, you may wanna consider making that clear in naming (i.e. third_party/kind_vcluster).

**Review by vladimirvivien — COMMENTED** [review]
This is a great start. I left some comments.

File: third_party/vcluster/vcluster.go
Comment (crandles) [reply third_party/vcluster/vcluster.go:126]: No, this doesn't require kind, this should work in any host k8s cluster.

I set it up via kind in the example so that it could presumably execute successfully (doubting there's a real cluster to access otherwise within CI).

**Comment by crandles** [comment]
golangci appears to be passing now for thew new code I've added, though errirng for some existing LOC:

```
third_party/helm/helm.go:240:9: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
		err = fmt.Errorf(missingHelm)
		      ^
third_party/flux/flux_setup.go:47:16: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
			return ctx, fmt.Errorf(NoFluxInstallationFoundMsg)
			            ^
third_party/flux/flux_setup.go:61:16: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
			return ctx, fmt.Errorf(NoFluxInstallationFoundMsg)
			            ^
pkg/env/env.go:229:11: printf: non-constant format string in call to (*testing.common).Skipf (govet)
		t.Skipf(message)
		        ^
pkg/env/env.go:513:22: printf: non-constant format string in call to (*testing.common).Skipf (govet)
					internalT.Skipf(message)
```

Assuming that won't block this PR.

**Comment by vladimirvivien** [comment]
@harshanarayana and @cpanato any idea why golangci failing on previously working code ?

**Comment by crandles** [comment]
errors were primarily for:
> SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)

It's possible that I'm using a different toolchain version, but as the warning seemed fair, I've fixed it via https://github.com/kubernetes-sigs/e2e-framework/pull/450/commits/cf9cbf28dfc7a407ed901b2da26b31ef4aabcc35

**Comment by vladimirvivien** [comment]
@crandles is this ready to go now ?

**Comment by crandles** [comment]
Yes, I believe so

**Comment by crandles** [comment]
rebased

**Review by vladimirvivien — APPROVED** [review]
LGTM

**Comment by vladimirvivien** [comment]
@harshanarayana @ShwethaKumbla @cpanato PTAL I think we need a second reviewer.

**Comment by harshanarayana** [comment]
I will check by eod tomorrow.

**Comment by crandles** [comment]
Seems to fail installing vcluster; can't reproduce. Any tips? I think I'm installing vcluster the same way other tools are (ko|kwok|kind)

**Comment by harshanarayana** [comment]
> Seems to fail installing vcluster; can't reproduce. Any tips? I think I'm installing vcluster the same way other tools are (ko|kwok|kind)

Helm and flux are installed slightly in a different way, https://github.com/kubernetes-sigs/e2e-framework/blob/777cbcd59fbbe572b71dde89c8f7b0b1a5ddc4a7/hack/install-helm.sh

```bash
test: install-helm install-flux ## Runs golang unit tests
	./hack/test-go.sh

**Comment by crandles** [comment]
Is there a recommended way to simulate the prow test runner locally?

This looks relevant: https://docs.prow.k8s.io/docs/build-test-update/#running-a-prowjob-locally, I'll try to follow it.

**Comment by crandles** [comment]
found a couple of issues:
- I was mistaken on the vcluster installation path, it is vclusterctl not vcluster
  - Apparently I've been falling back to an alternatively installed vcluster in my CI
- there was a flaw in the version defaulting logic

**Review by cpanato — APPROVED** [review]
thanks

/lgtm
/approve

**Comment by crandles** [comment]
Thanks for the reviews!

I will try to rebase next time, and I may follow up with an issue about how to run the pr tests locally. I need to spend more time looking at the pj on kind script.

**Pull request merged**
